### ISSUE TITLE
Destructuring "rest" pattern

### ISSUE THREAD

**Issue by ishehadeh** [janet-lang/janet#912]
Basically just varargs for destructuring.

When destructuringan array or tuple it can be useful to only bind the first few values and gathering the rest.
The syntax could be the same as varargs, making `[x y & rest]` a valid destructuring pattern equivalent to:
```janet
(let [x (get rhs 0)
      y (get rhs 0)
      rest (slice rhs 2)]
  ...)
```

I've got a [proof of concept](https://github.com/ishehadeh/janet/commit/cb5af974a443ed44f95a8772745cef2428f70c89) that implements this syntax.

It's missing some things, though:
- Any error checking, at all, for the left or right hand side
- I haven't really checked if it has any implications in contexts other than a basic `let/def` destructuring a tuple.
- It would probably be useful to match elements at the end too, something like `[w x & rest y z]`. This does make the current syntax a little weird, though.

Thoughs? I'm totally willing to implement the feature, just looking for input on it before moving forward.

**Comment by bakpakin** [comment]
Looks pretty good to me - might be good to consider supporting this with match.

As for supporting elements at the end, I don't think it's a good idea and is not supported by current functionality in function arguments.

**Comment by eko234** [comment]
Hi I was thinking about some destructuring features that I like and might be useful, @bakpakin , what do you think about syntax for aliases on destructures and what about destructuring rest of keys like in js `{key, ...rest}`?

**Comment by andrewchambers** [comment]
I think it could be quite handy but is also a performance trap I think a lot of people will fall into.

**Comment by ishehadeh** [comment]
Yeah it's definitely pretty slow atm. Some kind of view type would be really useful in this case (and imo a lot of other situations). That way 'rest could just be a reference to the original array + a start index and length.

**Comment by andrewchambers** [comment]
It's not really possible without completely reworking the internals because afaik, views would have no way to seeing when an array has been shrunk which could introduce segfaults/out of bounds access.

**Comment by ishehadeh** [comment]
Its definitely a lot of added complexity, especially for a dynamic language since it might not be clear whether you're working with the original structure or a view.

Although, they are really important for performance and ergonomics in some cases. Aside from destructuring strings could really benefit IMO.

**Comment by bakpakin** [comment]
View types are not on the roadmap - you could actually implement one yourself as an abstract type but probably not something that will be built in.

**Comment by pepe** [comment]
May I ask out of curiosity if this has any chance of adding to the language? As I already said a couple of times, I would appreciate this feature.

Yet I can understand @andrewchambers worries and support this attitude.

**Comment by ishehadeh** [comment]
Went ahead and cleaned up my implementation a bit, and added support in match.

**Comment by pepe** [comment]
I have installed this version on my work machine and tried with the biggest project I developed. I am not using the `& rest` pattern in the code, but it seems to work without any problem.

I will find a place to use the pattern in my code and report back.

**Comment by pyrmont** [comment]
This can be closed following the merge of PR #913.
